plugins/tc/tbf/main.go

```diff
@@ -0,0 +1,208 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"crypto/sha1"
+	"encoding/json"
+	"fmt"
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/containernetworking/cni/pkg/version"
+	"github.com/vishvananda/netlink"
+)
+
+type PluginConf struct {
+	types.NetConf
+	RuntimeConfig *struct{} `json:"runtimeConfig"`
+
+	// This is the previous result, when called in the context of a chained
+	// plugin. Because this plugin supports multiple versions, we'll have to
+	// parse this in two passes. If your plugin is not chained, this can be
+	// removed (though you may wish to error if a non-chainable plugin is
+	// chained.
+	// If you need to modify the result before returning it, you will need
+	// to actually convert it to a concrete versioned struct.
+	RawPrevResult *map[string]interface{} `json:"prevResult"`
+	PrevResult    *current.Result         `json:"-"`
+
+	// Add plugin-specifc flags here
+	Rate  int `json:"rate"`  //Bandwidth rate in Kbps for traffic through container. 0 for no limit. If rate is set, burst must also be set
+	Burst int `json:"burst"` //Bandwidth burst in Kb for traffic through container. 0 for no limit. If burst is set, rate must also be set
+}
+
+// parseConfig parses the supplied configuration (and prevResult) from stdin.
+func parseConfig(stdin []byte) (*PluginConf, error) {
+	conf := PluginConf{}
+
+	if err := json.Unmarshal(stdin, &conf); err != nil {
+		return nil, fmt.Errorf("failed to parse network configuration: %v", err)
+	}
+
+	// Parse previous result. Remove this if your plugin is not chained.
+	if conf.RawPrevResult != nil {
+		resultBytes, err := json.Marshal(conf.RawPrevResult)
+		if err != nil {
+			return nil, fmt.Errorf("could not serialize prevResult: %v", err)
+		}
+		res, err := version.NewResult(conf.CNIVersion, resultBytes)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse prevResult: %v", err)
+		}
+		conf.RawPrevResult = nil
+		conf.PrevResult, err = current.NewResultFromResult(res)
+		if err != nil {
+			return nil, fmt.Errorf("could not convert result to current version: %v", err)
+		}
+	}
+	// End previous result parsing
+
+	switch {
+	case conf.Burst < 0 || conf.Rate < 0:
+		return nil, fmt.Errorf("rate and burst must be a positive integer")
+	case conf.Burst == 0 && conf.Rate != 0:
+		return nil, fmt.Errorf("if rate is set, burst must also be set")
+	case conf.Rate == 0 && conf.Burst != 0:
+		return nil, fmt.Errorf("if burst is set, rate must also be set")
+	}
+
+	return &conf, nil
+
+}
+
+func getIfbDeviceName(networkName string, containerId string) (string, error) {
+	hash := sha1.New()
+	_, err := hash.Write([]byte(networkName + containerId))
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%x", hash.Sum(nil))[:15], nil
```

Comment:
I think this is ok, but I'd like other @containernetworking/cni-maintainers to weigh in.

Reply:
I think I'd prefer a short prefix (<= 4 characters).  That should still leave enough entropy (>=2^44) even just using the basic hex encoding.